### ISSUE TITLE
Duplication glitch (Smeltery output)

### ISSUE THREAD

**Issue by rob5251** [SlimeKnights/TinkersConstruct#2933]
Found a bug that causes lava to be used instead of the molten metal in the smeltery. 

When you go to empty molten metal (into a casting basin, or a cast in the casting table) via a faucet; if you have lava outputting to the casting basin/table it will cause the lava to be used in place of the metal that is suppose to be outputting. The faucet will appear to still be pouring out while this happens. Once the cast is complete you can extract it and the proper amount will flow out after.

Using this bug you can put one nugget of metal into the smeltery and get an entire block of metal out of it.

Steps to reproduce:

Step 1:
Have lava hooked up to the casting basin/casting table:
(Directly from ender tank or via pipes/conduits)
![step1](http://imgur.com/INuJhas.png)

Step 2:
Put in a single nugget of metal:
![step2](http://imgur.com/KPsOHc1.png)

Step 3:
Pour molten metal out into basin or cast via a faucet or conduit/pipe (it will appear as if its filling with lava, but the proper metal will be produced):
![step3](http://imgur.com/GVagSmy.png)
![step3](http://imgur.com/eJatZWP.png)

Step 4:
Get an entire block or ingot of metal from just one nugget...
![step4](http://imgur.com/zE3C11A.png)

Edit:
Did some further testing and if you use fluid conduits to output from the smeltery automatically directly into a casting basin you can potentially get several blocks of metal from this bug.
(Was able to get 37 blocks from one block of metal in one test)

**Comment by KnightMiner** [comment]
Does this happen with Tinkers faucets only, or with just Tinkers Construct? Sounds like another mod is messing with things.

**Comment by rob5251** [comment]
You can use faucets or pipes from other mods. The only requirement is that you have a way to have lava inputting into the casting table or basin.

**Comment by KnightMiner** [comment]
I mean does this work if you input lava using our faucets? If so, does this work with just Tinkers Construct installed?

**Comment by rob5251** [comment]
It can't be done with faucets alone. It requires lava being supplied via a pipe while you pour molten metal from the smeltery. When you go to pour the metal out of the smeltery the pipe will input lava at the same time and take the metals place (but still makes the metal block/casted item).

**Comment by jezzen** [comment]
Was thinking that it may be related to another annoying issue where the accepted liquid of the casting table and basin would be set when there is no valid recipe for it. After which, presumably because there is no extraction event, the setting won't be removed unless you set a valid recipe and add then remove fluid, or replace the block.

**Comment by bonii-xx** [comment]
There was lots of fixing regarding issues like this in 1.12. I'm going to assume it's fixed and/or related to how EIO inserts the liquids. If it's reproducable in 1.12 we'll probably find out once EIO is on 1.12. In a new issue.

**Comment by RunAge** [comment]
Still reproducable in 1.12 with EIO also with Mekanism https://streamable.com/w5kjs

**Comment by HenryLoenwind** [comment]
Looks to me like the basin sets the recipe without locking its tank or accepting liquid. So it sits around with a locked in recipe and an empty tank for a whole tick, or however long it takes for the faucet to insert liquid again.

BTW, Ender IO's conduit code is straight-forward:

```
    if (closestTank != null) {
      int filled = closestTank.tank.fill(toDrain.copy());

...

  public int fill(FluidStack resource) {
    return fluidHandler.fill(resource, true);
  }
```

**Comment by bonii-xx** [comment]
Indeed. I think it's because tanks (the logical class, not the blocks) have special handling when they are at 0 fluid. My guess is that it has the recipe set from the first insert, but is at 0 fluid amount, then the next insert comes along. When a recipe is set it just tries to do a regular tank fill, and because the amount is 0 it accepts it although it's a different fluid.
Anyway, added a few more fluid specific stuff in the fluid transfer, and I can't reproduce it with EIO anymore... let's see if it holds up. :D
